src/topology_cacher.py

```diff
@@ -0,0 +1,583 @@
+
+
+log = logging.getLogger(__name__)
+
+
+class DataError(Exception):
+    pass
+
+
+class ResourceInfo(namedtuple("ResourceInfo", "group_name name fqdn service_ids tags")):
+    SERVICE_ID_CE = "1"
+    SERVICE_ID_SCHEDD = "109"
+
+    def is_ce(self):
+        return self.SERVICE_ID_CE in self.service_ids
+
+    def is_schedd(self):
+        return self.SERVICE_ID_SCHEDD in self.service_ids
+
+    def to_dict(self) -> dict:
+        return dict(
+            group_name=self.group_name,
+            name=self.name,
+            fqdn=self.fqdn,
+            service_ids=self.service_ids,
+            tags=self.tags,
+        )
+
+
+class TopologyData:
+    def __init__(
+        self,
+        topology_base=TOPOLOGY,
+    ):
+        if "://" not in topology_base:
+            topology_base = "https://" + topology_base
+        self.topology_base = topology_base
+        self.resinfo_table = []
+        self.grouped_resinfo = {}
+        self.resinfo_by_name = {}
+        self.resinfo_by_fqdn = {}
+        self.projects = self.get_projects()
+        self.resources = self.get_resources()
+
+    def get_projects(self) -> ET.Element:
+        return self._get_data("/miscproject/xml")
+
+    def get_resources(self) -> ET.Element:
+        data = self._get_data("/rgsummary/xml")
+        self.resinfo_table = []
+        self.grouped_resinfo = {}
+        self.resinfo_by_name = {}
+        self.resinfo_by_fqdn = {}
+
+        #
+        # Build tables and indices for easier lookup
+        #
+        for eResourceGroup in data.findall("./ResourceGroup"):
+            group_name = safe_element_text(eResourceGroup.find("./GroupName"))
+            if not group_name:
+                log.warning(
+                    "Skipping malformed ResourceGroup: %s", elem2str(eResourceGroup)
+                )
+                continue
+            self.grouped_resinfo[group_name] = []
+
+            for eResource in eResourceGroup.findall("./Resources/Resource"):
+                resource_name = safe_element_text(eResource.find("./Name"))
+                fqdn = safe_element_text(eResource.find("./FQDN"))
+                service_ids = list(
+                    filter(
+                        None,
+                        [
+                            safe_element_text(svc)
+                            for svc in eResource.findall("./Services/Service/ID")
+                        ],
+                    )
+                )
+                if not resource_name or not fqdn or not service_ids:
+                    log.warning("Skipping malformed Resource: %s", elem2str(eResource))
+                    continue
+                tags = list(
+                    filter(
+                        None,
+                        [
+                            safe_element_text(tag)
+                            for tag in eResource.findall("./Tags/Tag")
+                        ],
+                    )
+                )
```

Comment:
since this is a pattern,
```
def findall_nonempty(elt, path):
    return list(filter(None, map(safe_element_text, elt.findall(path))))
```

and then

```suggestion
                tags = findall_nonempty(eResource, "./Tags/Tag")
```